File: Solution-Python/Longest_329.py
class Solution:
    # DFS + Memoization
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        self.m, self.n = len(matrix), len(matrix[0])
        cache = [[0]*self.n for i in range(self.m)]
        ans = 0
        for i in range(self.m):
            for j in range(self.n):
                ans = max(ans, self.dfs(matrix, i, j, cache))
        return ans
    
    def dfs(self, matrix: List[List[int]], i: int, j: int, cache: List[List[int]]) -> int:
        if cache[i][j] > 0:
            return cache[i][j]
        for d in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            x, y = i + d[0], j + d[1]
            if x >= 0 and y >= 0 and x < self.m and y < self.n and matrix[x][y] > matrix[i][j]:
                cache[i][j] = max(cache[i][j], self.dfs(matrix, x, y, cache))
        cache[i][j] += 1
        return cache[i][j]



    # Topological Sort
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        inDegree = [[0]*n for i in range(m)]
        for i in range(m):
            for j in range(n):
                for d in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                    x, y = i + d[0], j + d[1]
                    if x >= 0 and y >= 0 and x < m and y < n and matrix[x][y] > matrix[i][j]:
                        inDegree[x][y] += 1
        q = deque()
        for i in range(m):
            for j in range(n):
                if inDegree[i][j] == 0:
                    q.append((i, j))
        ans = 0
        while q:
            ans += 1
            for k in range(len(q)):
                i, j = q.popleft()
                for d in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                    x, y = i + d[0], j + d[1]
                    if x >= 0 and y >= 0 and x < m and y < n and matrix[x][y] > matrix[i][j]:
                        inDegree[x][y] -= 1
                        if inDegree[x][y] == 0:
                            q.append((x, y))
        return ans

File: Solution-Java/Longest_329.java
class Longest_329 {
    // Standard solution #2: DFS + Memoization
    // Time: O(m*n), Space(m*n)
    private static final int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    private int m, n;

    public int longestIncreasingPath(int[][] matrix) {
        m = matrix.length;
        n = matrix[0].length;
        int[][] cache = new int[m][n];
        int ans = 0;
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j)
                ans = Math.max(ans, dfs(matrix, i, j, cache));
        return ans;
    }

    private int dfs(int[][] matrix, int i, int j, int[][] cache) {
        if (cache[i][j] != 0) return cache[i][j];
        for (int[] d : dirs) {
            int x = i + d[0], y = j + d[1];
            if (0 <= x && x < m && 0 <= y && y < n && matrix[x][y] > matrix[i][j])
                cache[i][j] = Math.max(cache[i][j], dfs(matrix, x, y, cache));
        }
        return ++cache[i][j];
    }



    // Rewrite standard solution #3: Topological Sort
    // Time: O(m*n), Space(m*n)
    // Standard solution uses out-degree of each node to form Topological Sort from leaf to root
    // I use in-degree of each node to form Topological Sort from root to leaf
    private static final int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    private int m, n;

    public int longestIncreasingPath(int[][] matrix) {
        m = matrix.length;
        n = matrix[0].length;
        int[][] inDegree = new int[m][n];
        
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                for (int[] d : dirs) {
                    int x = i + d[0], y = j + d[1];
                    if (x >= 0 && x < m && y >= 0 && y < n && matrix[x][y] > matrix[i][j]) {
                        inDegree[x][y]++;
                    }
                }
            }
        }
        
        Deque<int[]> q = new ArrayDeque<>();
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (inDegree[i][j] == 0) {
                    q.add(new int[]{i, j});
                }
            }
        }
        
        int ans = 0;
        while (!q.isEmpty()) {
            ans++;
            for (int k = q.size(); k > 0; k--) {
                int[] node = q.removeFirst();
                for (int[] d : dirs) {
                    int x = node[0] + d[0], y = node[1] + d[1];
                    if (x >= 0 && x < m && y >= 0 && y < n && matrix[x][y] > matrix[node[0]][node[1]]) {
                        if (--inDegree[x][y] == 0) {
                            q.add(new int[]{x, y});
                        }
                    }
                }
            }
        }
        return ans;
    }
}